File: be/app/schemas/recovery.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.common import ApiModel


class RecoveryConstraints(ApiModel):
    allow_substitution: bool = False
    max_additional_delay_minutes: int | None = None


class RecoveryRequest(ApiModel):
    constraints: RecoveryConstraints | None = None


class RecoverySummary(ApiModel):
    risks_mitigated: int
    operational_changes: int
    recoverable_orders: int
    total_orders: int


class RecoveryAction(ApiModel):
    id: str
    what: str
    why: str
    expected_impact: str


class ManufacturingAction(RecoveryAction):
    product_id: str
    product_name: str
    baseline_quantity: int
    recovery_quantity: int
    change_quantity: int


class LogisticsAction(RecoveryAction):
    order_id: str
    original_warehouse_id: str
    original_warehouse_name: str
    recovery_warehouse_id: str
    recovery_warehouse_name: str
    vehicle_id: str
    baseline_route_id: str | None = None
    recovery_route_id: str | None = None
    baseline_eta_minutes: int | None = None
    recovery_eta_minutes: int | None = None
    baseline_flood_exposure: Literal["low", "medium", "high", "critical"] | None = None
    recovery_flood_exposure: Literal["low", "medium", "high", "critical"] | None = None
    action: Literal["reallocate-reroute", "reroute", "reallocate", "keep"]


class CommerceAllocation(ApiModel):
    product_id: str
    product_name: str
    quantity: int


class CommerceAction(RecoveryAction):
    order_id: str
    store_id: str
    store_name: str
    requested_product_id: str
    requested_product_name: str
    requested_quantity: int
    action: Literal["split-substitute", "delay", "fulfill", "fail"]
    allocations: list[CommerceAllocation]


class RecoveryResult(ApiModel):
    id: str
    simulation_id: str
    status: Literal["queued", "processing", "ready", "partial", "no-feasible-plan", "failed"]
    created_at: datetime
    completed_at: datetime | None = None
    summary: RecoverySummary | None = None
    manufacturing_actions: list[ManufacturingAction] | None = None
    logistics_actions: list[LogisticsAction] | None = None
    commerce_actions: list[CommerceAction] | None = None
    possible_next_actions: list[str] | None = None
    error: dict[str, str] | None = None

File: fe/src/domain/recovery.ts
import { z } from "zod";
import { apiErrorSchema, riskLevelSchema } from "./common";

const recommendationSchema = z.object({ what: z.string(), why: z.string(), expectedImpact: z.string() });
export const manufacturingActionSchema = recommendationSchema.extend({
  id: z.string(), productId: z.string(), productName: z.string(), baselineQuantity: z.number().nonnegative(), recoveryQuantity: z.number().nonnegative(), changeQuantity: z.number(),
});
export const logisticsActionSchema = recommendationSchema.extend({
  id: z.string(), orderId: z.string(), originalWarehouseId: z.string(), originalWarehouseName: z.string(), recoveryWarehouseId: z.string(), recoveryWarehouseName: z.string(), vehicleId: z.string(),
  baselineRouteId: z.string(), recoveryRouteId: z.string(), baselineEtaMinutes: z.number().nonnegative(), recoveryEtaMinutes: z.number().nonnegative(),
  baselineFloodExposure: riskLevelSchema, recoveryFloodExposure: riskLevelSchema, action: z.enum(["reallocate", "reroute", "reallocate-reroute"]),
});
export const commerceActionSchema = recommendationSchema.extend({
  id: z.string(), orderId: z.string(), storeId: z.string(), storeName: z.string(), requestedProductId: z.string(), requestedProductName: z.string(), requestedQuantity: z.number().positive(),
  action: z.enum(["fulfill", "split", "delay", "substitute", "prioritize", "split-substitute"]),
  allocations: z.array(z.object({ productId: z.string(), productName: z.string(), quantity: z.number().nonnegative() })),
});

const recoveryBaseSchema = z.object({ id: z.string(), simulationId: z.string(), createdAt: z.iso.datetime({ offset: true }) });
const recoveryPendingSchema = recoveryBaseSchema.extend({ status: z.enum(["queued", "processing"]) });
const recoveryFailedSchema = recoveryBaseSchema.extend({ status: z.literal("failed"), error: apiErrorSchema });
const recoveryResultSchema = recoveryBaseSchema.extend({
  status: z.enum(["ready", "partial", "no-feasible-plan"]), completedAt: z.iso.datetime({ offset: true }),
  summary: z.object({ risksMitigated: z.number().int().nonnegative(), operationalChanges: z.number().int().nonnegative(), recoverableOrders: z.number().int().nonnegative(), totalOrders: z.number().int().nonnegative() }),
  manufacturingActions: z.array(manufacturingActionSchema), logisticsActions: z.array(logisticsActionSchema), commerceActions: z.array(commerceActionSchema),
  possibleNextActions: z.array(z.string()),
});
export const recoveryPlanSchema = z.discriminatedUnion("status", [recoveryPendingSchema, recoveryFailedSchema, recoveryResultSchema]);
export const recoveryGenerationRequestSchema = z.object({
  constraints: z.object({ allowSubstitution: z.boolean().optional(), maxAdditionalDelayMinutes: z.number().nonnegative().optional() }).optional(),
});

export type ManufacturingAction = z.infer<typeof manufacturingActionSchema>;
export type LogisticsAction = z.infer<typeof logisticsActionSchema>;
export type CommerceAction = z.infer<typeof commerceActionSchema>;
export type RecoveryPlan = z.infer<typeof recoveryPlanSchema>;
